src/policy/marksweepspace/native_ms/block.rs

```diff
@@ -0,0 +1,462 @@
+// adapted from Immix
+
+use atomic::Ordering;
+
+use super::BlockList;
+use super::MarkSweepSpace;
+use crate::util::heap::chunk_map::*;
+use crate::util::linear_scan::Region;
+use crate::vm::ObjectModel;
+use crate::{
+    util::{metadata::side_metadata::SideMetadataSpec, Address, OpaquePointer, VMThread},
+    vm::VMBinding,
+};
+
+use std::num::NonZeroUsize;
+
+/// A 64KB region for MiMalloc.
+/// This is also known as MiMalloc page. We try to avoid getting confused with the OS 4K page. So we call it block.
+/// This is a non-zero block, and `Option<Block>` can still a word size value.
+// TODO: What if we actually use the first block?
+#[derive(Debug, Clone, Copy, PartialOrd, PartialEq)]
+#[repr(C)]
+pub struct Block(NonZeroUsize);
+
+impl From<Address> for Block {
+    #[inline(always)]
+    fn from(address: Address) -> Block {
+        debug_assert!(address.is_aligned_to(Self::BYTES));
+        debug_assert!(!address.is_zero());
+        Self(unsafe { NonZeroUsize::new_unchecked(address.as_usize()) })
+    }
+}
```

Comment:
We expect to turn an aligned address to a region:
https://github.com/mmtk/mmtk-core/blob/06144875903b36b432587eb7462e0272750b71d9/src/util/linear_scan.rs#L79

We do expect it to always succeed, and we do not expect to handle the error case in any circumstance. I feel it is unnecessary to use `TryFrom/Into`, which is designed for 'conversions that may fail in a controlled way under some circumstances'. I don't think that is our case. For us, if the assertion fails, that means we have a bug in the code that we should fix.

Reply:
If the purpose is to use `Block` as a `Region` and that requires `From<Address>`, we can keep this `From<Address>` implementation as is for now.

But I think there is a deeper problem in `Region`, because if a bug-free program only ever converts aligned addresses to `Region`, then `Region::is_aligned` should always return true (an it should be unnecessary).  It is similar to the `Block::is_zero` function and the unnecessary assertions.

I don't fully agree with the use of `From` instead of `TryFrom`.  In the standard library, for example, `i64` implements `From<i32>`, but `i32` doesn't implement `From<i64>`.  If a piece of code converts `i64` to `i32`, but "the i64 here should always fit into i32 unless there is a bug", it still doesn't make it a valid use case of `impl From<i64> for i32`.  It needs to use `try_from` because *the Rust language* cannot protect the programmer from supplying an out-of-range value.

Likewise, in our program, an `Address` can express any address, aligned or not, zero or not.  The Rust language cannot prevent the programmer from converting an unaligned or zero `Address` to `Block`.  I think `Block::try_from(xxx).unwrap()` and `Block::try_from(xxx).expect("message")` also mean "if this fails, it is a bug, and we should fix it".

Reply:
I think the key point of `TryFrom/TryInto` is error handling. We do not need error handling in this case. We could simply hard fail with a panic if the input address is invalid, and fix the code that causes the issue. 

We could require `Region` to implement both: `trait Region: From<Address> + TryFrom<Address>`. But we would always use `Block::from(Block::align(addr))`, rather than `Block::try_from(addr).unwrap_or_else(|| Block::from(Block::align(addr)))`. In that case, why should we implement `TryFrom`?

Reply:
What about adding a method `unsafe fn from_address_unchecked(addr: Address) -> Self` to the `Region` trait, and not require `From<Address>` or `TryFrom<Address>`?  We can document it so that it is a contract that the caller must ensure `addr` is aligned and non-null.  This can distinguish it from `From::from` which "never fails" and `TryFrom::try_from` which "may fail in a controlled way".

Reply:
Yeah. That sounds reasonable. Rust says `From` should never fail, and that means we should not have `From<Address>` for `Region` at all.

Reply:
I submitted a PR for this: https://github.com/mmtk/mmtk-core/pull/695.